### ISSUE TITLE
amp-story: Incorrect focus order/handling

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#32511]
# Incorrect focus order/handling

## WCAG Level
Level A

## Priority
Critical

## Pages/screens/components affected
* [Navigation](https://www.gstatic.com/amphtml/stamp/qa/navigation.html)
* [RTL](https://www.gstatic.com/amphtml/stamp/qa/rtl.html)
* [Video](https://www.gstatic.com/amphtml/stamp/qa/video.html)
* [Audio - amp-story background audio](https://www.gstatic.com/amphtml/stamp/qa/audio-story.html)
* [Audio - amp-story-page background audio](https://www.gstatic.com/amphtml/stamp/qa/audio-page.html)
* [Audio - amp-story using amp-audio](https://www.gstatic.com/amphtml/stamp/qa/amp-audio.html)
* [Animation](https://www.gstatic.com/amphtml/stamp/qa/animation.html)
* [Sharing](https://www.gstatic.com/amphtml/stamp/qa/sharing.html)
* [Tooltip](https://www.gstatic.com/amphtml/stamp/qa/tooltip.html)
* [Page attachment](https://www.gstatic.com/amphtml/stamp/qa/page-attachment.html)
* [Branching](https://www.gstatic.com/amphtml/stamp/qa/branching.html#page=page2)
* [Consent](https://www.gstatic.com/amphtml/stamp/qa/consent.html)
* [Prerender - Image](https://www.gstatic.com/amphtml/stamp/qa/prerender-img.html?amp_js_v=0.1#origin=https://www.google.com&prerenderSize=1&visibilityState=prerender&paddingTop=0&p2r=0&horizontalScrolling=0&csi=1&aoh=15526927198732&viewerUrl=https://www.google.com&history=1&storage=1&cid=1&cap=swipe,navigateTo,cid,fragment,replaceUrl)
* [Prerender - Video](https://www.gstatic.com/amphtml/stamp/qa/prerender-video.html?amp_js_v=0.1#origin=https://www.google.com&prerenderSize=1&visibilityState=prerender&paddingTop=0&p2r=0&horizontalScrolling=0&csi=1&aoh=15526927198732&viewerUrl=https://www.google.com&history=1&storage=1&cid=1&cap=swipe,navigateTo,cid,fragment,replaceUrl)
* [AMP Story Player - Player](https://www.gstatic.com/amphtml/stamp/qa/player.html)
* [AMP Story Player - AMP Story 360](https://www.gstatic.com/amphtml/stamp/qa/amp-story-360.html)

## Description
For people who navigate using a keyboard or keyboard-like device (such as a switch control or a joystick) the focus order of interactive controls in a document should be logical and intuitive.

Throughout all of the tested AMP stories, we noted issues with focus handling - ranging from issues with the general controls for all stories, and issues specific to particular stories' content.

## User impact
If the focus order is not logical, sighted keyboard users are likely to be confused and irritated when navigating through a site, particularly if the focus appears to "jump" in an illogical manner to different, seemingly unconnected parts of the screen. This issue will be even more problematic of users of assistive technologies, who may not have the benefit of being able to perceive the overall layout/structure of the page to realise that that the illogical focus order has potentially moved them to a completely separate part of the current page.

When focus is not programmatically moved to the most appropriate location following an interaction, users will be left to manually navigate again to the most appropriate location in the page. For sighted keyboard users, this can be an inconvenience/irritation, particularly if focus was lost/reset to the start of the document and they have to cycle through the entire page again. Blind and low vision assistive technology users may be unaware of what change actually occurred on the page as a result of their interaction. After performing an action, these users may be confused/left wondering if their action was actually successful. In cases where focus is lost (generally caused by the element that previously had focus being disabled or removed from the page), some browsers may error-correct and attempt to relocate the user's focus to the nearest focusable element, or they may simply reset the focus to the start of the document. Overall, it will likely make the user's experience of navigating the page unnecessarily tedious and confusing.

## Required solution
Make sure that the order in which elements receive keyboard focus follows a meaningful and logical sequence. The order does not necessarily have to follow the order implied by the visual layout, as long as it is logical. Nonetheless, the focus should ideally not "jump" erratically between different parts of the layout.

Pay attention to source order. Sequencing elements in the code so they follow a logical order helps to provide a logical and meaningful focus order.

In the case of elements that are visually hidden, make sure that focus does not move to these interactive elements.

This solution must be applied to all instances of the issue identified within the test sample, then applied to all other instances of the same issue identified throughout the rest of the website.

### Focus lost when activating controls

The generic play/pause and sound on/off controls used across all stories are currently implemented as two separate buttons (with one button visible, and one button set to `display:none`). When the user activates one control, that control is then hidden and the other control is revealed. As a result, the currently focused element disappears, which can lead to focus being lost/reset to the start of the document (though browsers tend to error-correct for this).

![Illustration of the focus behaviour for the play/pause control - once the focused play control is activated, it is hidden and swapped for the pause control, but focus is not placed on the pause control](https://user-images.githubusercontent.com/54933393/107251336-26621700-6a2c-11eb-955a-eab434745982.png)

```
<div class="i-amphtml-paused-display">
  <button class="i-amphtml-story-pause-control i-amphtml-story-button" aria-label="Pause story"></button>
  <button class="i-amphtml-story-play-control i-amphtml-story-button" aria-label="Play story"></button>
</div>
```

In this case, we recommend programmatically setting focus to the button that gets shown/replaces the previously focused and activated control. Alternatively, consider reworking the way these toggle controls work by only using a single control and dynamically changing its graphical image and `aria-label`, which would obviate the need for focus handling.

### Focus remains on navigation buttons when activated
When activating the previous/next/replay buttons, focus remains on the button that the user just activated. As a result, a keyboard user (and, in particular, users of assistive technologies such as screen readers) will have to navigate backwards to reach the content of the newly shown page.

We recommend moving focus programmatically to the start of the page every time navigation occurs.

### Redundant focus stops
In [AMP Story Player - Player](https://www.gstatic.com/amphtml/stamp/qa/player.html) there appear to be two "Share" buttons (as the player presumably loads separate stories, all with their own separate controls). This result in a confusing/redundant navigation experience (and, currently, setting focus to the second share button causes the layout to visually break).

![The resulting broken layout when navigating to the second share button](https://user-images.githubusercontent.com/54933393/107251441-3ed23180-6a2c-11eb-910a-8617dca732e4.png)

If technically feasible, we'd recommend eliminating the redundant focus stop and making sure that only a single share control is present in the story overall, even if it's constructed from separate sub-stories.

### Controls in other pages receiving focus
When navigating using `Tab`/`Shift`+`Tab`, it is possible to set focus to elements in currently non-active/dimmed pages of the AMP story.

![An AMP story, showing a particular page (and other pages dimmed), but with focus indication around a link in a subsequent dimmed page](https://user-images.githubusercontent.com/54933393/107251498-4e517a80-6a2c-11eb-9476-f7989ee945a4.png)

![An AMP story, showing a particular page (and other pages dimmed), but with focus indication around a link in a subsequent dimmed page](https://user-images.githubusercontent.com/54933393/107251570-64f7d180-6a2c-11eb-8bd5-14653801d9b8.png)

![An AMP story, showing a particular page (and other pages dimmed), but with focus indication around a link in a subsequent dimmed page(https://user-images.githubusercontent.com/54933393/107251616-717c2a00-6a2c-11eb-81bd-c107d8784104.png)

Note that the content of any non-active/dimmed pages *is* currently hidden from assistive technologies using `aria-hidden="true"`. As a result, an assistive technology user navigating with `Tab`/`Shift`+`Tab` will focus onto "nothing" - no announcement will be made by their AT on what element just received focus, making the experience even more confusing.

The solution to these issues is to explicitly make focusable elements (buttons, links, etc.) in non-active/dimmed pages non-focusable by programmatically adding `tabindex="-1"` (and removing this attribute when the page becomes active/visible).

### Invisible elements receiving focus

In some of the samples, we found that focus goes to invisible elements which should only be available when opening a panel or modal dialog. In most cases, this also leads to broken layouts - see the separate issue *Pages visually break when navigating with keyboard*.

As an example, in [Page attachment](https://www.gstatic.com/amphtml/stamp/qa/page-attachment.html), the link to `www.google.com`, which should only be active when the user "swiped up" to open the attachment panel, still receives focus even when the panel is closed.

![Illustration of the google link receiving focus when the 'swipe up' panel is still closed](https://user-images.githubusercontent.com/54933393/107251664-835dcd00-6a2c-11eb-97ec-35a5f91c708d.png)

![Illustration of the google link receiving focus when the 'swipe up' panel is still closed](https://user-images.githubusercontent.com/54933393/107251713-92447f80-6a2c-11eb-8358-24687a1962e7.png)

Another example of this behaviour can be seen in [Sharing](https://www.gstatic.com/amphtml/stamp/qa/sharing.html), where the sharing dialog, as well as the "bookend" screen used in small-screen/smartphone view, have focusable `<amp-social-share>` controls. These controls take focus even when the user hasn't opened the sharing dialog or reached the bookend screen (which also leads to broken layouts - again, see the separate issue *Pages visually break when navigating with keyboard*).

![The sharing and bookend screen showing at the same time, as they contain focusable elements, despite not being explicitly opened](https://user-images.githubusercontent.com/54933393/107251937-c91a9580-6a2c-11eb-9fa6-1f5a4bb46e55.png)

The solution to these issues is to explicitly make focusable elements (buttons, links, etc.) that are currently not visible (inside panels or modal dialogs that are closed) non-focusable by programmatically adding `tabindex="-1"` (and removing this attribute when the panel or modal become visible).

### Modal dialogs/panels

As noted in the previous section, focusable elements inside modal dialogs (such as the sharing dialog) and panels ("Swipe up" page attachment, bookend screen in small-screen/smartphone view) currently receive focus even when the dialogs/panels are not open. However, there are further focus-related issues with these components.

When the sharing modal dialog is opened, focus remains on the "Share" button, rather than being moved to the dialog itself. Users have to navigate through the (now dimmed) page to reach the dialog. When the sharing modal dialog is closed, focus is not programmatically moved back to the control that originally opened the dialog.

![After opening the sharing dialog, focus is visibly still on the 'Share' button](https://user-images.githubusercontent.com/54933393/107252012-db94cf00-6a2c-11eb-9b83-87ae013cecab.png)

When modal dialogs (such as the sharing dialog, as well as the non-native consent dialog in [Consent](https://www.gstatic.com/amphtml/stamp/qa/consent.html)) are open, focus is not maintained inside the dialog. Users are able to `Tab`/`Shift`+`Tab` through the underlying page, and even trigger controls (i.e. a user can open the "Share" dialog, and still move to the previous/next slide behind the semi-transparent overlay; or, when the non-native consent dialog is shown, users can still navigate to the "Share" button and trigger *that* dialog/panel as well).

![With the sharing dialog open, the user's focus is visibly on the 'Replay' button in the underlying page](https://user-images.githubusercontent.com/54933393/107252072-e7809100-6a2c-11eb-8bd4-56c76d8f7826.png)

![With the web-based consent dialog open, the user navigated to the 'share' button in the underlying page](https://user-images.githubusercontent.com/54933393/107252139-f5cead00-6a2c-11eb-8e81-ac5c6b574029.png)

![With the web-based consent dialog open, the user navigated to the 'share' button and triggered the sharing panel/dialog](https://user-images.githubusercontent.com/54933393/107252186-00894200-6a2d-11eb-8e4b-1ae5f9450861.png)

![With the web-based consent dialog open, the user navigated to the 'next' button in the underlying page in desktop view](https://user-images.githubusercontent.com/54933393/107252239-0bdc6d80-6a2d-11eb-9502-2c495307640e.png)

![With the bookend panel/modal open, a user was able to also trigger the 'Share' dialog/panel](https://user-images.githubusercontent.com/54933393/107252285-1ac32000-6a2d-11eb-8d08-b975b65cbaa9.png)

The solution to these issues is to explicitly move focus to the modal/panel when it is opened, and to make any focusable elements (buttons, links, etc.) that are in the underlying page (behind the panel/modal dialog) non-focusable by programmatically adding `tabindex="-1"` (and removing this attribute when the panel or modal become visible).

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied:
1. Use `Tab`/`Shift`+`Tab` to navigate through the page.
2. Verify that the order in which elements receive focus is logical and consistent.
3. Verify that if an element is visually hidden, it does not receive focus.
4. For modal dialogs or panels, verify that focus is moved to the widget once it's opened, and maintained inside the widget (i.e. that focus does not go back to the underlying page).

## Related standards
* [WCAG 2.1 Success Criterion 2.4.3 Focus Order (Level A)](https://www.w3.org/TR/WCAG21/#focus-order)

## More information
* [W3C - Placing the interactive elements in an order that follows sequences and relationships within the content](https://www.w3.org/WAI/WCAG21/Techniques/general/G59)
* [The A11Y Project - How-to: Hide content](https://www.a11yproject.com/posts/2013-01-11-how-to-hide-content/)

## Test data
Test date: January 2021

**Comment by gmajoulet** [comment]
Closing as the keyboard navigation experience has been largely tested by our team after many bug fixes.
